File: SignalServiceKit/src/Messages/OWSMessageUtils.m
//
//  Copyright (c) 2021 Open Whisper Systems. All rights reserved.
//

#import "OWSMessageUtils.h"
#import "AppContext.h"
#import "MIMETypeUtil.h"
#import "MessageSender.h"
#import "TSAccountManager.h"
#import "TSAttachment.h"
#import "TSAttachmentStream.h"
#import "TSIncomingMessage.h"
#import "TSMessage.h"
#import "TSOutgoingMessage.h"
#import "TSQuotedMessage.h"
#import "TSThread.h"
#import "UIImage+OWS.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@implementation OWSMessageUtils

+ (NSUInteger)unreadMessagesCount
{
    __block NSUInteger numberOfItems;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        numberOfItems = [InteractionFinder unreadCountInAllThreadsWithTransaction:transaction.unwrapGrdbRead];
    }];

    return numberOfItems;
}

+ (NSUInteger)unreadMessagesCountExcept:(TSThread *)thread
{
    __block NSUInteger numberOfItems;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        NSUInteger allCount = [InteractionFinder unreadCountInAllThreadsWithTransaction:transaction.unwrapGrdbRead];
        InteractionFinder *interactionFinder = [[InteractionFinder alloc] initWithThreadUniqueId:thread.uniqueId];
        NSUInteger threadCount = [interactionFinder unreadCountWithTransaction:transaction.unwrapGrdbRead];
        numberOfItems = (allCount - threadCount);
    }];

    return numberOfItems;
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Messages/OWSMessageUtils.swift
//
//  Copyright (c) 2020 Open Whisper Systems. All rights reserved.
//

import Foundation
import SignalCoreKit

extension OWSMessageUtils {
    static public func unreadMessageCountPromise() -> Promise<UInt> {
        databaseStorage.read(.promise) { readTx in
            InteractionFinder.unreadCountInAllThreads(transaction: readTx.unwrapGrdbRead)
        }
    }

    @objc
    static public func updateApplicationBadgeCount() {
        let fetchBadgeCount = { () -> Promise<UInt> in
            // The main app gets to perform this synchronously
            if CurrentAppContext().isMainApp {
                return .value(unreadMessagesCount())
            } else {
                return self.unreadMessageCountPromise()
            }
        }


        fetchBadgeCount().done {
            CurrentAppContext().setMainAppBadgeNumber(Int($0))
        }.catch { error in
            owsFailDebug("Failed to update badge number: \(error)")
        }
    }
}
